airbyte-cdk/python/airbyte_cdk/sources/concurrent_source/thread_pool_manager.py

```diff
@@ -2,81 +2,86 @@
-        self._prune_futures(self._futures)
         self._futures.append(self._threadpool.submit(function, *args))
 
     def _prune_futures(self, futures: List[Future[Any]]) -> None:
         """
         Take a list in input and remove the futures that are completed. If a future has an exception, it'll raise and kill the stream
         operation.
 
-        Pruning this list safely relies on the assumptions that only the main thread can modify the list of futures.
+        We are using a lock here as without it, the algorithm would not be thread safe
         """
-        if len(futures) < self._max_concurrent_tasks:
-            return
+        with self._lock:
+            if len(futures) < self._max_concurrent_tasks:
+                return
 
-        for index in reversed(range(len(futures))):
-            future = futures[index]
+            for index in reversed(range(len(futures))):
+                future = futures[index]
 
-            if future.done():
-                # Only call future.exception() if the future is known to be done because it will block until the future is done.
-                # See https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Future.exception
-                optional_exception = future.exception()
-                if optional_exception:
-                    exception = RuntimeError(f"Failed reading with error: {optional_exception}")
-                    self._stop_and_raise_exception(exception)
-                futures.pop(index)
+                if future.done():
+                    # Only call future.exception() if the future is known to be done because it will block until the future is done.
+                    # See https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Future.exception
+                    optional_exception = future.exception()
+                    if optional_exception:
+                        # Exception handling should be done in the main thread. Hence, we only store the exception and expect the main
+                        # thread to call raise_if_exception
+                        self._most_recently_seen_exception = RuntimeError(f"Failed reading with error: {optional_exception}")
+                    futures.pop(index)
 
     def shutdown(self) -> None:
         self._threadpool.shutdown(wait=False, cancel_futures=True)
 
     def is_done(self) -> bool:
         return all([f.done() for f in self._futures])
 
+    def raise_if_exception(self) -> None:
+        if self._most_recently_seen_exception:
+            self._stop_and_raise_exception(self._most_recently_seen_exception)
+
     def check_for_errors_and_shutdown(self) -> None:
         """
         Check if any of the futures have an exception, and raise it if so. If all futures are done, shutdown the threadpool.
         If the futures are not done, raise an exception.
         :return:
         """
+        self.raise_if_exception()
```

Comment:
won't this prevent the threadpool from shutting down?

maybe we can call self.shutdown() then `raise_if_exception()`

Reply:
`raise_if_exception` calls `_stop_and_raise_exception` underneath. I think it means the method name could be improve. What do you think about `shutdown_if_exception` and we can still raise the exception as part of this method